### PR TITLE
test: snapshots for content tests

### LINKED ISSUE #246: Add snapshot testing.
Now that we have implemented [`testza`](https://github.com/MarvinJWendt/testza), we should optimize the testing process. An important step is, to introduce Snapshot testing.
We can create a Snapshot for every `Printer`, to validate its correct output.

### PR DESCRIPTION
**Pull request by TobiaszCudnik**
### Description
Snapshots for various content-related tests (35):
- `Render()`
- `GetContent()`
- `AssertContains()`

Not covered:
- `...NilPrint()`
- `test?Print??Contains()`

Also skipped for now are tests without assertions, eg `TestAreaPrinter_RemoveWhenDone`. It may be a good idea to cover them as well.

**TODO** - Depends on [fixed testza snapshots](https://github.com/MarvinJWendt/testza/pull/65). Currently theres a local module replace in `go.mod`, which has to be removed before merging. 

- [ ] Bug Fix
- [ ] New Feature
- [ ] Documentation
- [x] Other

### Related Issue
Fixes #246

### To-Do Checklist
- [x] I tested my changes
- [ ] **N/A** I have commented every method that I created/changed
- [ ] **N/A** I updated the examples to fit with my changes
- [ ] **N/A** I have added tests for my newly created methods

### CODE REVIEW

**Comment by MarvinJWendt** [comment]
Hi @TobiaszCudnik, the fix in testza will be released today, thanks 😉 
To the snapshot testing:
I don't really know, if we should snapshot variable contents (assertions) like in: 

```go
func TestBoxPrinter_WithTitleWithTitleTopCenter(t *testing.T) {
	p := pterm.BoxPrinter{}
	p2 := p.WithTitleTopCenter().WithTitle("a").Sprint("Lorem Ipsum")

	testza.AssertContains(t, p2, "Lorem Ipsum")
	testza.SnapshotCreateOrValidate(t, t.Name(), p2)
}
```

I think Snapshots only make sense when complicated terminal output has to be checked (as for terminal output with color codes in particular). In the example above, the name assertion can be hardcoded (`testza.AssertContains(t, p2, "Lorem Ipsum")`), so I don't think it needs a snapshot. Or do you have something in mind that I don't see? 😄 

I would say we snapshot every terminal output, like `Render()` and Print functions, as they are hard to assert with the color codes. Everything that can be tested via simple assertions (eg.: variable content) does not need a snapshot IMO.

Thanks for working on this :)

**Comment by TobiaszCudnik** [comment]
Changes:
- removed snapshots for `AssertContains` 
- bumped `testza` to v0.2.12 from upstream
- added snapshots for every step of the demo

The motivation behind snapshotting `AssertContains` was that it doesnt assert the whole output (like `AssertEquals`), but the snapshots would. On the other hand, snapshots are costly to maintain.

I agree that snapshots have the biggest advantage with complicated output, thats why I reused the demo code and created an acceptance test, which has a snapshot on each step.

Lesson learned: date generation doesnt work well with snapshots. Neither do random numbers.

Unfortunately, [BarChart seems to have OS-dependent output](https://github.com/pterm/pterm/runs/4162449675?check_suite_focus=true) and fails on the CI, eg `TestBarChartPrinter_RenderHorizonzal`. Do you have an idea why?

> Feel free to ask questions anytime and thanks for working on this :)

No problem ;)

**Comment by MarvinJWendt** [comment]
Sadly I can't tell you why the BarChart has different output. I'll have to investigate that :)
For now you could name the test something like `[test name] + "_" + runtime.GOOS`, so that each platform has their own snapshot of the BarChart printer. But that's just a workaround, I'll take a look why the output differs.

**Comment by MarvinJWendt** [comment]
Hey @TobiaszCudnik, you might want to merge the current master branch into yours. The blank lines when running the tests, are now omitted, so the output looks cleaner.

**Comment by TobiaszCudnik** [comment]
Ive merged the master and regenerated the snapshots, but I see that even on the new ubuntu CI youve added the output is still different in some cases (barchart and csv table).

Its always shorter on the CI, but Ive checked it on macos and all snapshots pass. Maybe its related to the terminal being "dumb" on the CI, maybe not 256 color?

**Comment by MarvinJWendt** [comment]
My strong guess is that the detected terminal size for ubuntu is different. PTerm tries to fit the output to the terminal size. If the size cannot be detected, it uses a default value of 80 characters length. The Windows output for the BarChart is 80 chars long while the one from ubuntu is 60 chars long. You could try to set the terminal size manually to 80 inside the tests, or in the test initialization. I guess that would fix the problem. Let me know if it helps.

**Comment by TobiaszCudnik** [comment]
I've commented out those tests with a note and created #296 to track it.

This PR is ready to be merged, as other snapshots seem to work just fine.

**Comment by TobiaszCudnik** [comment]
It indeed was the size detection, I have a [PR](https://github.com/pterm/pterm/pull/297) with hardcoded values which passes.

I will introduce a new API to force-set the dimensions.

**Comment by MarvinJWendt** [comment]
> I will introduce a new API to force-set the dimensions.

Do you want to make the API public or private for testing?

**Comment by MarvinJWendt** [comment]
Hi @TobiaszCudnik, I see that you use `Srender()` to get the content from the printers, and save this to the snapshots. Do you think it would make more sense to capture the output of the previous `Render()` call, and validate its output? If someone messes up the `Render()` function, the end-user won't get the correct output, and it won't be caught by the unit tests.

**Comment by TobiaszCudnik** [comment]
@MarvinJWendt This was a lot more shaving than I initially expected :) Terminal size was hardcoded on modules' inits.

Changes:
- SetForcedTerminalSize()
  - tests
- CI tests fixed
- optimized Render() & Srender() tests
- NewDefaultBarChart() constructor
    - updated docs & examples
- OS dependent-snapshots (some)
    - in-repo only for linux
- NewDefaultParagraph()
    - updated docs & examples

>  I see that you use Srender() to get the content from the printers, and save this to the snapshots. Do you think it would make more sense to capture the output of the previous Render() call, and validate its output?

Initially I didnt want to alter existing tests in any way, but I do agree that it should be optimized.

**Comment by MarvinJWendt** [comment]
Woah, that's a lot! I really like the work, but you introduced a breaking change by removing `pterm.DefaultBarChart`, which also doesn't fit to the "paradigm" of PTerm to access every printer with `DefaultXxx`. Am I seeing this right, that you created a constructor, so that the terminal size gets reloaded when a new printer is created? I totally see the intention, but since the bar chart is widely used in other projects, it might not be an option to introduce a breaking change there (some bigger projects would not build anymore, if they update PTerm). Also, a problem is that you cannot modify the Default values of the BarChart anymore. Normally, a user can overwrite values of `DefaultXxx` to change the printer settings on a project level. With a constructor, this is no longer possible. See: https://pterm.sh/#/docs/customizing/override-default-printer
Especially for the `BarChart`, a user might want to change the actual bar character of the default printer.

I see the difficulty in refreshing the terminal size on the printers and why it's important, tho.

I guess we could introduce another option to the `BarChart` and `Paragraph`, like `FitToTerminalSize bool`. If that value is true, the size get's updated before every re-render to fit the terminal. If that value is false, the size will be static. What do you think about that? If you have a better idea, tell me :)

---

BTW: Huge thanks for all your work, this really helps a lot 😄

**Comment by TobiaszCudnik** [comment]
Changes:
- reverted constructors
- added RecalculateTerminalSize

Valid point regarding breaking APIs, especially in a test-related PR. I guess a proper resize support is out of scope here, but having a consistent policy regarding terminal size would be nice. Currently some values are hardcoded and some calculated during a render (eg grep `GetTerminalWidth()`).

For now Ive, reverted all the changes to the existing APIs and simply added `RecalculateTerminalSize()`, which may also be useful for the end user. Although its manual, it will adapt new instances after a resize (which is exactly what was needed for the tests).

File: barchart.go [hunk 1]
```diff
@@ -32,11 +32,21 @@ var (
...
+// GetDefaultBarChartWidth calculates the width of DefaultBarChart.
+func GetDefaultBarChartWidth() int {
+	return GetTerminalHeight() * 2 / 3
+}
+
+// GetDefaultBarChartHeight calculates the height of DefaultBarChart.
+func GetDefaultBarChartHeight() int {
+	return GetTerminalWidth() * 2 / 3
+}
+
```
Comment (MarvinJWendt) [inline barchart.go:49]: I think we should remove those methods and inline them for now, until we have a solid terminal rezising approach :)

File: pterm.go [hunk 2]
```diff
@@ -57,3 +57,10 @@ func DisableStyling() {
 	RawOutput = true
 	DisableColor()
 }
+
+// RecalculateTerminalSize updates already initialized terminal dimensions. Has to be called after a termina resize to guarantee proper rendering. Applies only to new instances.
+func RecalculateTerminalSize() {
+	DefaultBarChart.Width = GetDefaultBarChartWidth()
```
Comment (MarvinJWendt) [inline pterm.go:63]:
```suggestion
	DefaultBarChart.Width = GetTerminalWidth() * 2 / 3
```

File: barchart.go (hunk 1 above)
Comment (MarvinJWendt) [inline barchart.go:49]: I think we should remove those methods and inline them for now, until we have a solid terminal resizing approach :)

File: pterm.go [hunk 3]
```diff
@@ -57,3 +57,10 @@ func DisableStyling() {
...
+	DefaultBarChart.Height = GetDefaultBarChartHeight()
```
Comment (MarvinJWendt) [inline pterm.go:64]:
```suggestion
	DefaultBarChart.Height = GetTerminalHeight() * 2 / 3
```

File: testdata/snapshots/TestDemo0.testza [hunk 4]
```diff
@@ -0,0 +1 @@
+(string) ""
```
Comment (MarvinJWendt) [inline testdata/snapshots/TestDemo0.testza:1]: This is an empty snapshot, which is probably committed by accident?

File: testdata/snapshots/TestDemo1.testza [hunk 5]
```diff
@@ -0,0 +1 @@
+(string) ""
```
Comment (MarvinJWendt) [inline testdata/snapshots/TestDemo1.testza:1]: This is an empty snapshot, which is probably committed by accident?

File: testdata/snapshots/TestDemo2.testza [hunk 6]
```diff
@@ -0,0 +1 @@
+(string) ""
```
Comment (MarvinJWendt) [inline testdata/snapshots/TestDemo2.testza:1]: This is an empty snapshot, which is probably committed by accident?

File: testdata/snapshots/TestDemo3.testza [hunk 7]
```diff
@@ -0,0 +1 @@
+(string) ""
```
Comment (MarvinJWendt) [inline testdata/snapshots/TestDemo3.testza:1]: This is an empty snapshot, which is probably committed by accident?

File: testdata/snapshots/TestDemo4.testza [hunk 8]
```diff
@@ -0,0 +1 @@
+(string) ""
```
Comment (MarvinJWendt) [inline testdata/snapshots/TestDemo4.testza:1]: This is an empty snapshot, which is probably committed by accident?

File: testdata/snapshots/TestDemo5.testza [hunk 9]
```diff
@@ -0,0 +1 @@
+(string) ""
```
Comment (MarvinJWendt) [inline testdata/snapshots/TestDemo5.testza:1]: This is an empty snapshot, which is probably committed by accident?

File: testdata/snapshots/TestDemo6.testza [hunk 10]
```diff
@@ -0,0 +1 @@
+(string) ""
```
Comment (MarvinJWendt) [inline testdata/snapshots/TestDemo6.testza:1]: This is an empty snapshot, which is probably committed by accident?

File: testdata/snapshots/TestDemo7.testza [hunk 11]
```diff
@@ -0,0 +1 @@
+(string) ""
```
Comment (MarvinJWendt) [inline testdata/snapshots/TestDemo7.testza:1]: This is an empty snapshot, which is probably committed by accident?

File: testdata/snapshots/TestDemo8.testza [hunk 12]
```diff
@@ -0,0 +1 @@
+(string) ""
```
Comment (MarvinJWendt) [inline testdata/snapshots/TestDemo8.testza:1]: This is an empty snapshot, which is probably committed by accident?

**Review by MarvinJWendt — CHANGES_REQUESTED** [review]
I have commented some files, other than that, the PR is ready in my eyes :)

File: pterm.go (hunk 2 above)
Comment (TobiaszCudnik) [reply pterm.go]: Reverted with a comment.

File: pterm.go (hunk 3 above)
Comment (TobiaszCudnik) [reply pterm.go]: Reverted with a comment.

File: barchart.go (hunk 1 above)
Comment (TobiaszCudnik) [reply barchart.go]: Reverted with a comment.

File: testdata/snapshots/TestDemo0.testza (hunk 4 above)
Comment (TobiaszCudnik) [reply testdata/snapshots/TestDemo0.testza]: Good catch, there was a `setup` call missing. Demo snapshots were working when run directly, but not with other tests.

**Review by MarvinJWendt — APPROVED** [review]
LGTM! Thanks for the contribution 🚀

**Pull request merged**